Fix view snap back after lock-on in mp1

Unreal-Dan opened the pull request:
This continuously writes out the viewangle pitch to match the lock-on pitch even while locked on. 

This ensures that the instant the lock-on is released the view will not be a snap back to the old position, but instead it will remain exactly where the lock-on was last aiming.

I've only implemented it for mp1 wii edition, I can do it for the other versions but that will take a little longer. 

I also don't have the gamecube versions of mp1 or mp2, but I suppose I could acquire them.

Note this tends to the issue here: https://github.com/shiiion/dolphin/issues/28

Also excuse me, I'm a git noob, the commit author is my desktop user even though I set my global git config email and username -- I'm not sure what the hell is going on there but if anybody could tell me how to fix that I would do so.

Antidote commented on Source/Core/Core/PrimeHack/Mods/FpsControls.cpp:
```diff
@@ -1060,6 +1081,8 @@ namespace prime {
       mp1_static.tweak_player_address = 0x804ddff8;
       mp1_static.cplayer_address = 0x804d3c20;
+      mp1_static.object_list_ptr_address = 0x804bfc30;
+      mp1_static.camera_uid_address = 0x804c4a08;
```
I don't think these addresses are static.

Unreal-Dan replied:
I just stole them from the noclip code, quite literally copy pasted their name and everything. 

If that's the case then there might be some fixing in the noclip code to do as well

SirMangler replied:
Our rule of thumb is generally anything around `804bxxxx` or lower is static, we've had no reports or issues so far with these addresses so far.

SirMangler commented on Source/Core/Core/PrimeHack/Mods/FpsControls.cpp:
```diff
@@ -136,17 +145,29 @@ namespace prime {
-    calculate_pitch_delta();
     writef32(pitch, mp1_static.pitch_address);
     writef32(pitch, mp1_static.pitch_goal_address);
+    if (locked) {
```
Assuming pitch will be identical to what the game has currently stored, I would move the pitch writes to the `else` block since that is the only case it should be written. Ideally you should return before writing pitch when locked if it is going to be writing the exact same value anyway.

Unreal-Dan replied:
Unfortunately if you only write the pitch while not locked then you get this 1-frame snap where your view snaps back to the previous location.

This happens because the frame where you release lock is processed and rendered before this mod is run, which means by the time we are able to start writing the pitch again the game has already rendered a frame without locking and in turn causes a 1-frame stutter-like snap immediately after locking.

One solution would be to have a better hookpoint to run this mod, one where the input has been processed but the frame hasn't been rendered yet.

Short of a better hook the only way to handle the 1-frame snap is to just write out the pitch throughout the entirety of being locked on, even if it's not actually visible while locked on. That way the very frame the lock is released it will render the correct viewangle.

SirMangler commented on Source/Core/Core/PrimeHack/Mods/FpsControls.cpp:
```diff
@@ -136,17 +145,29 @@ namespace prime {
+    if (locked) {
       write32(0, mp1_static.yaw_vel_address);
-      return;
+      // Calculate the pitch based on the XF matrix of player so that we always
+      // write out the pitch even while locked onto the target, this prevents
+      // the view from snapping back after releasing a lock-on
```
I would place this comment before the `calculate_pitch_locked` function is defined, rather than in the run function, to try keep things clean.

SirMangler commented on Source/Core/Core/PrimeHack/Mods/FpsControls.cpp:
```diff
@@ -74,6 +74,15 @@ namespace prime {
     pitch = std::clamp(pitch, -1.52f, 1.52f);
   }
 
+  void FpsControls::calculate_pitch_locked() {
```
This really should handle regions and games. I understand you only support MP1 Trilogy with this PR, however you haven't defined static addresses for MP1 PAL, and this function will run with the expectation that those static addresses are defined.

SirMangler reviewed (commented):
Generally decent solution. In a perfect world you would do the same for calculating `yaw`, and run these two calculations when handling the cursor. That would solve many other problems like with grapple and Prime 3's motion segments.

Unreal-Dan commented:
> Generally decent solution. In a perfect world you would do the same for calculating yaw, and run these two calculations when handling the cursor. That would solve many other problems like with grapple and Prime 3's motion segments.

I don't believe the yaw is necessary for MP1, it seems to work perfectly with just the pitch because the game is already controlling the yaw and only the pitch is being overwritten by us.

But you say that would solve issues with grapple and prime 3's motion segments? I haven't had any issues with grapple in MP1 so far with this fix, and I haven't gotten to MP3 yet so I'll have to get back to you unless you can demo what you mean.

Unreal-Dan commented:
I've pushed a new branch, fixed the commit author, fixed the comment you pointed out in a bad place, added the PAL address for those two (but I can't test it, I just stole the address from noclip, should work).

As for the one open issue about where to write the pitch, I believe we have to write out the pitch 100% of the time now for this to work.

SirMangler commented:
> 
> I haven't gotten to MP3 yet so I'll have to get back to you unless you can demo what you mean.

The issues with the camera snapping are more prevalent in MP3, largely with grapple lasso. You're quite right, I forgot we only write yaw **velocity** and not directly to the value.

Good work with this.
